packages/dataviews/src/types.ts

```diff
@@ -499,6 +498,9 @@ export interface ViewBaseProps< Item > {
 	setOpenedFilter: ( fieldId: string ) => void;
 	view: View;
 	density: number;
+	onPinItem?: ( itemId: string ) => void;
+	onUnpinItem?: ( itemId: string ) => void;
+	pinnedItems: string[];
```

Comment:
I wonder if a separate `pinnedItems` key is better than a callback `isPinned( item: Item ) => boolean`? I'm not sure just asking to make sure we discuss the right API.

Reply:
Yeah, I also [shared some thoughts](https://github.com/WordPress/gutenberg/pull/65629#discussion_r1776890673) about this. While reviewing I brainstormed a few approaches:

1. At the DataViews component level: `pinnedItems` prop, like this PR proposes. This is also what we do with `selection` (as a list of ids).
2. At the Fields API level, have a `isPinned` callback. I struggled to find a good place for this: perhaps this callback could be part just of the field "id", but I don't like that it's special.
3. At the Fields API level: consumers should provide a field whose id is `is-pinned` and whose `getValue` would return whether or not the field is pinned. That way we wouldn't need an `isPinned` callback. After giving it a thought, it doesn't make much sense to me.
4. The View API. The view is about the UI state, but I couldn't come up with anything I liked because you'd have to store item's id there (e.g.; `view.pinnedItems: [ xxx, yyy ]`.

Ultimately, my thinking is that pinning is like selection. Both are about making some "rows" (items/records) special, and they are not about "columns" (fields) or UI state (view). That's why the approach in this PR sounds the most promising to me.

Reply:
Nice summary. For me, I think it's best to be at the component level like here but was wondering if it should be a function or an array like selection. The current approach works for me.